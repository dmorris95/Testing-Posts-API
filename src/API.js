const jsonApiUrl = 'https://jsonplaceholder.typicode.com';

export const fetchPosts = async () => {
    const response = await fetch(`${jsonApiUrl}/posts`);
    if (!response.ok) {
        throw new Error('Network response was not good');
    }
    return response.json();
};

export const createPost = async (post) => {
    const response = await fetch(`${jsonApiUrl}/posts`, {
        method: 'POST',
        body: JSON.stringify(post),
        headers: {
            'Content-type': 'application/json; charset=UTF-8',
        },
    });
    if (!response.ok) {
        throw new Error('Failed to add post');
    }
    return response.json();
};

export const updatePost = async (post) => {
    const response = await fetch(`${jsonApiUrl}/posts/${post.id}`, {
        method: 'PUT',
        body: JSON.stringify(post),
        headers: {
            'Content-type': 'application/json; charset=UTF-8',
        },
    });
    if (!response.ok) {
        throw new Error('Failed to update post');
    }
    return response.json();
};

export const deletePost = async (postId) => {
    const response = await fetch(`${jsonApiUrl}/posts/${postId}`, {
        method: 'DELETE',
    });
    if (!response.ok) {
        throw new Error('Failed to delete post');
    }
    return response.json();
};