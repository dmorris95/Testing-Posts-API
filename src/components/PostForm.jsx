import { useMutation, useQueryClient } from "@tanstack/react-query";
import { useCallback, useEffect, useMemo, useState } from "react";
import { Form, Col, Button, Alert } from "react-bootstrap";
import { updatePost, createPost } from "../API";
import React from "react";

const PostForm = ({ post, resetSelectedPost }) => {
    //useState to populate form fields if a post was selected for update
    const [title, setTitle] = useState('');
    const [body, setBody] = useState('');
    const [userId, setUserId] = useState('');
    
    const queryClient = useQueryClient();
    const [showSuccessAlert, setShowSuccessAlert] = useState(false);

    useEffect(() => {
        if (post) {
            setTitle(post.title);
            setUserId(post.userId);
            setBody(post.body);
        } else{
            setTitle('');
            setBody('');
            setUserId('');
        }
    }, [post]);

    const { mutate, isLoading, isError, error } = useMutation({
        mutationFn: post ? updatePost : createPost,
        onSuccess: (data) => {
            setShowSuccessAlert(true);
            if (post) {
                console.log('Post updated successfully');
                queryClient.setQueryData(['posts'], (existingPosts) => {
                    return existingPosts.map((post) => (post.id === data.id ? data : post));
                })
            } else {
                console.log('Post created with ID:', data.id);
                queryClient.setQueryData(['posts'], (existingPosts) => [...existingPosts, data]);
            }
            //queryClient.invalidateQueries(['posts']);
            setTimeout(() => setShowSuccessAlert(false), 1000);
            setBody('');
            setTitle('');
            setUserId('');
            setTimeout(() => resetSelectedPost(), 1000);
        },
    });

    const newPost = useMemo(() => ({
        id: post?.id,
        userId: userId,
        title: title,
        body: body
    }), [post, userId, title, body]);

    const handleSubmit = useCallback((event) => {
        event.preventDefault();
        mutate(newPost);
    }, [newPost, mutate]);

    return (
        <div>
            {isError && <Alert variant="danger">An error occurred: {error.message}</Alert>}
            {showSuccessAlert && <Alert variant="success">Post successfully {post ? 'updated' : 'created'}</Alert>}
            <Col className="m-2 md-6">
                <Form onSubmit={handleSubmit} className="border p-3 rounded-4 border-dark">
                    <h3>Post Form</h3>
                    <Form.Group className="mb-3" controlId="title">
                        <Form.Label>Title</Form.Label>
                        <Form.Control type="text" placeholder="Enter Title" name="title" value={title} onChange={(e) => setTitle(e.target.value)} required />
                    </Form.Group>
                    <Form.Group className="mb-3" controlId="userid">
                        <Form.Label>UserID</Form.Label>
                        <Form.Control type="number" placeholder="Enter UserID" name="userid" value={userId} onChange={(e) => setUserId(e.target.value)} required />
                    </Form.Group>
                    <Form.Group className="mb-3" controlId="body">
                        <Form.Label>Body</Form.Label>
                        <Form.Control name="body" as="textarea" rows={3} value={body} onChange={(e) => setBody(e.target.value)} required />
                    </Form.Group>
                    {post && <Button variant="primary" type="submit" disabled={isLoading}>{isLoading ? 'Updating...' : 'Update Post'}</Button>}
                    {!post && <Button variant="primary" type="submit" disabled={isLoading}>{isLoading ? 'Adding...' : 'Add Post'}</Button>}
                </Form>
            </Col>
        </div>
    );
}

export default PostForm