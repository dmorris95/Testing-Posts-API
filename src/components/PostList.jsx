import { useQueryClient, useMutation, useQuery } from "@tanstack/react-query";
import { useCallback, useMemo, useState } from "react";
import { Card, Button, Spinner, Alert, Form } from "react-bootstrap";
import { fetchPosts, deletePost } from "../API";
import React from "react";

const PostList = ({ onPostClick }) => {
    const queryClient = useQueryClient();
    const [filterId, setFilterId] = useState('');

    const { data, isLoading, isSuccess, error } = useQuery({
        queryKey: ['posts'],
        queryFn: fetchPosts,
        refetchOnReconnect: true,
        refetchOnWindowFocus: true,
        retry: 3,
        retryDelay: attemptIndex => Math.min(1000 * 2 ** attemptIndex, 30000),
        staleTime: 5 * 60 * 1000, 
        cacheTime: 15 * 60 * 1000
    });

    const delPost = useMutation({
        mutationFn: deletePost,
        onSuccess: (data, postId) => {
            console.log('Post deleted successfully')
            queryClient.setQueryData(['posts'], (existingPosts) => {
                return existingPosts.filter(post => post.id !== postId);
            });
        },
    });

    //Post Memoization
    const filterPosts = useMemo(() => {
        //Data check in case an undefined value is returned from the API call
        return data ? data.filter(post => post.userId.toString().includes(filterId)) : [];
    }, [data, filterId]);

    //Memoize the postClick (The UserSelect in this program)
    const handlePostClick = useCallback((post) => {
        onPostClick(post);
    }, [onPostClick]);


    if (isLoading) return <Spinner animation="border" role="status"><span className="visually-hidden">Loading...</span></Spinner>;
    if (error) return <Alert variant="danger">Error fetching data</Alert>;

    return (
        <div>
            <input className="mb-3 bg-light border-primary rounded-2" name="filterId" type="number" value={filterId} placeholder="Filter by User ID" onChange={(e) => setFilterId(e.target.value)}/>
            {isSuccess && filterPosts.map(post => (
                <Card key={post.id} className="mb-3">
                    <Card.Body>
                        <Card.Title>{post.title}</Card.Title>
                        <Card.Text>{post.body}</Card.Text>
                        <Button variant="primary" onClick={() => handlePostClick(post)}>Edit</Button>
                        <Button 
                            variant="danger" 
                            data-testid={`del-button-${post.id}`} 
                            onClick={() => delPost.mutate(post.id)}
                        >
                        Delete
                        </Button>
                    </Card.Body>
                </Card>
            ))}
        </div>
    );
}

export default PostList