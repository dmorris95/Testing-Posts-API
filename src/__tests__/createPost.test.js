import React from "react";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import '@testing-library/jest-dom';
import PostForm from '../components/PostForm';
import { createPost, fetchPosts } from "../API";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import PostList from "../components/PostList";

//Mock CreatePost from the API.js
jest.mock('../API', () => ({
    createPost: jest.fn(),
    fetchPosts: jest.fn()
}));

const queryClient = new QueryClient();
const renderUsingClient = (ui) => {
    return render(
    <QueryClientProvider client={queryClient}>
        {ui}
    </QueryClientProvider>
    );
};

describe('createPost Function', () => {
    test('submit form and calls createPost from API', async () => {
        const mockPost = [{ id: 1, title: 'Mock Post', body: 'Mock Body', userId: 1 }];
        const newPost = { id: 2, title: 'Created Post', body: 'Created body', userId: 2};

        //simulate initial post
        fetchPosts.mockResolvedValue(mockPost);
        //simulate adding post
        createPost.mockResolvedValue(newPost);

        const { asFragment } = renderUsingClient(
            <>
                <PostForm />
                <PostList onPostClick={jest.fn()} />
            </>
        );

        fireEvent.change(screen.getByLabelText(/Title/i), { target: { value: 'Created Post'}});
        fireEvent.change(screen.getByLabelText(/UserID/i), { target: { value: '2'}});
        fireEvent.change(screen.getByLabelText(/Body/i), { target: { value: 'Created body'}});
        //Simulate adding the Post via Add Post button
        fireEvent.click(screen.getByText(/Add Post/i));
        
        //Ensure UI was updated correctly
        await waitFor(() => {
            expect(screen.getByText('Created Post')).toBeInTheDocument();
            expect(screen.getByText('Created body')).toBeInTheDocument();
        });
        //Confirm createPost was called
        expect(createPost).toHaveBeenCalledWith({
            title: 'Created Post',
            userId: '2',
            body: 'Created body'
        });

        //Test Snapshot
        expect(asFragment()).toMatchSnapshot();
    });  
});