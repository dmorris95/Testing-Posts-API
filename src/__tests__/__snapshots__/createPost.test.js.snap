// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPost Function submit form and calls createPost from API 1`] = `
<DocumentFragment>
  <div>
    <div
      class="fade alert alert-success show"
      role="alert"
    >
      Post successfully created
    </div>
    <div
      class="m-2 md-6 col"
    >
      <form
        class="border p-3 rounded-4 border-dark"
      >
        <h3>
          Post Form
        </h3>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="title"
          >
            Title
          </label>
          <input
            class="form-control"
            id="title"
            name="title"
            placeholder="Enter Title"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="userid"
          >
            UserID
          </label>
          <input
            class="form-control"
            id="userid"
            name="userid"
            placeholder="Enter UserID"
            required=""
            type="number"
            value=""
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="body"
          >
            Body
          </label>
          <textarea
            class="form-control"
            id="body"
            name="body"
            required=""
            rows="3"
          />
        </div>
        <button
          class="btn btn-primary"
          type="submit"
        >
          Add Post
        </button>
      </form>
    </div>
  </div>
  <div>
    <input
      class="mb-3 bg-light border-primary rounded-2"
      name="filterId"
      placeholder="Filter by User ID"
      type="number"
      value=""
    />
    <div
      class="mb-3 card"
    >
      <div
        class="card-body"
      >
        <div
          class="card-title h5"
        >
          Mock Post
        </div>
        <p
          class="card-text"
        >
          Mock Body
        </p>
        <button
          class="btn btn-primary"
          type="button"
        >
          Edit
        </button>
        <button
          class="btn btn-danger"
          data-testid="del-button-1"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      class="mb-3 card"
    >
      <div
        class="card-body"
      >
        <div
          class="card-title h5"
        >
          Created Post
        </div>
        <p
          class="card-text"
        >
          Created body
        </p>
        <button
          class="btn btn-primary"
          type="button"
        >
          Edit
        </button>
        <button
          class="btn btn-danger"
          data-testid="del-button-2"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
